### ISSUE TITLE
pymysql.err.OperationalError: (1045, "Access denied for user 'root'@'172.19.0.6' (using password: YES)")

### ISSUE THREAD

**Issue by Natetronn** [frappe/frappe_docker#448]
## Description of the issue

My issue is similar to this one, although on production (that issue is over a year old, didn't want to necro open it, and it was noted to open new issue for production; also, not using bench, as this is production so, not all is relevant; although I did try some of the solutions there, without success ): https://github.com/frappe/frappe_docker/issues/239

## Steps to reproduce the issue

1. After what appeared to be a successful setup of Docker et al, following the [Docker installation instructions](https://github.com/frappe/bench#docker-installation)
2. Get to the point of starting things up by running the following command:
```
docker exec -it \                          
    -e "SITE_NAME=$SITE_NAME" \
    -e "DB_ROOT_USER=$DB_ROOT_USER" \
    -e "MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD" \
    -e "ADMIN_PASSWORD=$ADMIN_PASSWORD" \
    -e "INSTALL_APPS=erpnext" \
    mysitecom_erpnext-python_1 docker-entrypoint.sh new
```

### Observed result
Access denied for user roor error:
`pymysql.err.OperationalError: (1045, "Access denied for user 'root'@'172.19.0.6' (using password: YES)")`

### Expected result
Expected root to have access to mariadb

### Stacktrace / full error message if available

```
docker exec -it \                          
    -e "SITE_NAME=$SITE_NAME" \
    -e "DB_ROOT_USER=$DB_ROOT_USER" \
    -e "MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD" \
    -e "ADMIN_PASSWORD=$ADMIN_PASSWORD" \
    -e "INSTALL_APPS=erpnext" \
    mysitecom_erpnext-python_1 docker-entrypoint.sh new

Attempt 1 to connect to mariadb:3306
Attempt 1 to connect to redis-queue:6379
Attempt 1 to connect to redis-cache:6379
Attempt 1 to connect to redis-socketio:6379
Connections OK

Traceback (most recent call last):
  File "/home/frappe/frappe-bench/commands/new.py", line 118, in <module>
    main()
  File "/home/frappe/frappe-bench/commands/new.py", line 70, in main
    db_port=db_port,
  File "/home/frappe/frappe-bench/apps/frappe/frappe/installer.py", line 72, in _new_site
    no_mariadb_socket=no_mariadb_socket,
  File "/home/frappe/frappe-bench/apps/frappe/frappe/installer.py", line 106, in install_db
    setup_database(force, source_sql, verbose, no_mariadb_socket)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/__init__.py", line 16, in setup_database
    return frappe.database.mariadb.setup_db.setup_database(force, source_sql, verbose, no_mariadb_socket=no_mariadb_socket)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/mariadb/setup_db.py", line 39, in setup_database
    if force or (db_name not in dbman.get_database_list()):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/db_manager.py", line 60, in get_database_list
    return [d[0] for d in self.db.sql("SHOW DATABASES")]
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 121, in sql
    self.connect()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 74, in connect
    self._conn = self.get_connection()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/mariadb/database.py", line 90, in get_connection
    local_infile = frappe.conf.local_infile)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/__init__.py", line 94, in Connect
    return Connection(*args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 325, in __init__
    self.connect()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 599, in connect
    self._request_authentication()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 861, in _request_authentication
    auth_packet = self._read_packet()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 684, in _read_packet
    packet.check_error()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/protocol.py", line 220, in check_error
    err.raise_mysql_exception(self._data)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/err.py", line 109, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1045, "Access denied for user 'root'@'172.19.0.6' (using password: YES)")
```

Note: I'm able to access maria db like so: `docker exec -ti mysitecom_mariadb_1 bash` and then `mysql -u root -p` and enter the root password set in .env to access the DBs:

```
MariaDB [(none)]> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| _123-redacted-123  |
| information_schema |
| mysql              |
| performance_schema |
+--------------------+
4 rows in set (0.001 sec)
```
Does the following look correct? Should root have more host access than localhost?

```
MariaDB [(none)]> SELECT user,authentication_string,plugin,host FROM mysql.user;
+-------------------+-----------------------+--------+------------+
| user              | authentication_string | plugin | host       |
+-------------------+-----------------------+--------+------------+
| root              |                       |        | localhost  |
| _38f74ec15f7433a1 |                       |        | 172.19.0.4 |
| _38f74ec15f7433a1 |                       |        | %          |
+-------------------+-----------------------+--------+------------+
3 rows in set (0.000 sec)
```

Any direction would be greatly appreciated, thanks!

**Comment by revant** [comment]
I'm getting following in mariadb:

```
❯ docker exec -it frappe_docker_mariadb_1 mysql -uroot -padmin
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 13
Server version: 10.3.28-MariaDB-1:10.3.28+maria~focal mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> select host, user from mysql.user;
+-----------+------+
| host      | user |
+-----------+------+
| %         | root |
| localhost | root |
+-----------+------+
2 rows in set (0.000 sec)
```

**Comment by Natetronn** [comment]
Thanks @revant !

I tried that. In fact, I tried a few different versions lol:

```
MariaDB [(none)]> SELECT user,host FROM mysql.user;
+-------------------+------------+
| user              | host       |
+-------------------+------------+
| root              | %          |
| _38f74ec15f7433a1 | 172.19.0.% |
| root              | 172.19.0.% |
| root              | localhost  |
+-------------------+------------+
4 rows in set (0.001 sec)
```
Same issue:

```
docker exec -it \                                               
    -e "SITE_NAME=$SITE_NAME" \
    -e "DB_ROOT_USER=$DB_ROOT_USER" \
    -e "MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD" \
    -e "ADMIN_PASSWORD=$ADMIN_PASSWORD" \
    -e "INSTALL_APPS=erpnext" \
    erpduerzacom_erpnext-python_1 docker-entrypoint.sh new
Attempt 1 to connect to mariadb:3306
Attempt 1 to connect to redis-queue:6379
Attempt 1 to connect to redis-cache:6379
Attempt 1 to connect to redis-socketio:6379
Connections OK
Traceback (most recent call last):
  File "/home/frappe/frappe-bench/commands/new.py", line 118, in <module>
    main()
  File "/home/frappe/frappe-bench/commands/new.py", line 70, in main
    db_port=db_port,
  File "/home/frappe/frappe-bench/apps/frappe/frappe/installer.py", line 72, in _new_site
    no_mariadb_socket=no_mariadb_socket,
  File "/home/frappe/frappe-bench/apps/frappe/frappe/installer.py", line 106, in install_db
    setup_database(force, source_sql, verbose, no_mariadb_socket)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/__init__.py", line 16, in setup_database
    return frappe.database.mariadb.setup_db.setup_database(force, source_sql, verbose, no_mariadb_socket=no_mariadb_socket)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/mariadb/setup_db.py", line 39, in setup_database
    if force or (db_name not in dbman.get_database_list()):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/db_manager.py", line 60, in get_database_list
    return [d[0] for d in self.db.sql("SHOW DATABASES")]
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 121, in sql
    self.connect()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 74, in connect
    self._conn = self.get_connection()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/mariadb/database.py", line 90, in get_connection
    local_infile = frappe.conf.local_infile)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/__init__.py", line 94, in Connect
    return Connection(*args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 325, in __init__
    self.connect()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 599, in connect
    self._request_authentication()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 861, in _request_authentication
    auth_packet = self._read_packet()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 684, in _read_packet
    packet.check_error()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/protocol.py", line 220, in check_error
    err.raise_mysql_exception(self._data)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/err.py", line 109, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1045, "Access denied for user 'root'@'172.19.0.2' (using password: YES)")
```

**Comment by revant** [comment]
Can you only debug mariadb with custom config?

only start mariadb container as separate project with custom config, try publishing ports to host, and try to access mysql shell from host machine.

then try to connect from separate container.

bitnami/mariadb allows % host for user by default. I've not manually set it anytime. I don't know why it's not allowing user access.

Can you confirm the root username and password? `echo $DB_ROOT_USER` and `echo $MYSQL_ROOT_PASSWORD`
may be they are going in blank?

**Comment by revant** [comment]
Tests seem to pass every day? I can't guess what's wrong.

**Comment by articlecat** [comment]
I'm seeing the same "access denied" errors when attempting to deploy ERPNext via Docker Compose using the [single bench install instructions](https://github.com/frappe/frappe_docker/blob/develop/docs/single-bench.md). I'm using ERPNext v13.0.1 and Frappe v13.0.2.

Here are the Docker Compose logs for the MariaDB container:
```
mariadb_1                 | 2021-04-12 11:43:22+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 1:10.3.28+maria~focal started.
mariadb_1                 | 2021-04-12 11:43:22+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mariadb_1                 | 2021-04-12 11:43:22+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 1:10.3.28+maria~focal started.
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] mysqld (mysqld 10.3.28-MariaDB-1:10.3.28+maria~focal) starting as process 1 ...
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Using Linux native AIO
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Uses event mutexes
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Number of pools: 1
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Using SSE2 crc32 instructions
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Completed initialization of buffer pool
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Creating shared tablespace for temporary tables
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Waiting for purge to start
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: 10.3.28 started; log sequence number 1625461; transaction id 20
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] Plugin 'FEEDBACK' is disabled.
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] InnoDB: Buffer pool(s) load completed at 210412 11:43:23
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] Server socket created on IP: '::'.
mariadb_1                 | 2021-04-12 11:43:23 0 [Warning] 'proxies_priv' entry '@% root@de0b5cd61d8b' ignored in --skip-name-resolve mode.
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] Reading of all Master_info entries succeeded
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] Added new Master_info '' to hash table
mariadb_1                 | 2021-04-12 11:43:23 0 [Note] mysqld: ready for connections.
mariadb_1                 | Version: '10.3.28-MariaDB-1:10.3.28+maria~focal'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
mariadb_1                 | 2021-04-12 11:43:35 14 [Warning] Access denied for user 'frappe'@'172.18.0.10' (using password: NO)
mariadb_1                 | 2021-04-12 11:43:35 15 [Warning] Access denied for user 'root'@'172.18.0.10' (using password: YES)
mariadb_1                 | 2021-04-12 11:44:35 16 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:44:35 17 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:44:46 18 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:44:46 19 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:45:35 20 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:45:35 21 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:46:35 22 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:46:35 23 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:47:25 24 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:47:25 25 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:47:30 26 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:47:30 27 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:47:35 28 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:47:35 29 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:48:35 30 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:48:35 31 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:48:54 32 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:48:54 33 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:49:35 34 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:49:35 35 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:50:36 36 [Warning] Access denied for user '_dc6c201278f92bf4'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:50:36 37 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.13' (using password: YES)
mariadb_1                 | 2021-04-12 11:51:02 38 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
mariadb_1                 | 2021-04-12 11:51:02 39 [Warning] Access denied for user '_4c6393feb2ccbcde'@'172.18.0.4' (using password: YES)
```

**Comment by cl-alessandro-molari** [comment]
@articlecat I'm having the same issue.

**Comment by revant** [comment]
I really have no idea how to solve this.

I still can't reproduce the issue.

**Comment by todfox** [comment]
I am having the same issue

**Comment by revant** [comment]
I do get warnings, they may occur when mariadb is starting up or creating the db and containers try to connect it.

```
2021-04-22  1:14:19 175 [Warning] Access denied for user 'frappe'@'10.0.1.4' (using password: NO)
```

Once all containers are running, site-creator is done creating sites, mariadb is configured with custom conf, everything starts working as expected on my end. Just tried in PWD again and it is working as expected. Tests also pass.

https://labs.play-with-docker.com/?stack=https://raw.githubusercontent.com/frappe/frappe_docker/develop/tests/pwd.yml

**Comment by articlecat** [comment]
Found the issue! When following the [single bench install instructions](https://github.com/frappe/frappe_docker/blob/develop/docs/single-bench.md), the `DB_ROOT_USER` variable must remain on the default value of `root` in the `.env` file. This is because the [mariadb image](https://hub.docker.com/_/mariadb) does not support changing the root username.

After clearing all of my volumes (with `docker volume prune`) and starting over with `DB_ROOT_USER=root` in `.env`, Frappe successfully installs ERPNext.

**Comment by Natetronn** [comment]
@revant thanks for looking into this! Unfortunately, I've since given up on using (installing) Frappe/ERPNext.

I'd be happy to discuss my reasons in greater detail but, on the other hand, I'm not sure how to give constructive criticism of an open source project like this. Not without it sounding like I'm just complaining; without contributing. 

That being said, I do want you to know that I appreciate you taking the time to look into this. And hopefully this was just a one off issue that I alone was having. I'll leave it open, however, as it appears others may be hitting this as well.

**Comment by revant** [comment]
@Natetronn.

I really don't know why there is this error? Is it slow machine that is not able to start and config mariadb before frappe/erpnext tries to connect to it? I may have never faced this because I use at least 2xCPU and 4GB RAM and PWD allocates enough resources during IST day time.

Can you try to use cli params to mariadb in its docker-compose/service section instead of using a mounted `.cnf` file. May be that will boot it faster? check this https://github.com/frappe/frappe_docker/issues/303#issuecomment-673766430

I think containerized install is easier than setting up whole os and dependencies manually from scratch. I'm ready to improve anything if I find some clue.

> I'd be happy to discuss my reasons in greater detail but, on the other hand, I'm not sure how to give constructive criticism of an open source project like this. Not without it sounding like I'm just complaining; without contributing.

You can complain! May be just create a new issue or forum post to point out things (many posts already exist). Try to keep it very objective. Try avoid rants, you can express it! Edit it if you feel it's too harsh. Point out things that can be improved.
Or just add to #353

I improve frappe_docker because I understand this repo completely. Other repos I can't help much, I just understand some parts of the code. I will try to help.

**Comment by revant** [comment]
root password is loaded as `os.environ.get()`. I don't know if it escapes special characters. Can you try without special characters in root password?

**Comment by imchivaa** [comment]
tried to use `MYSQL_ROOT_PASSWORD=$'H@lloWor!d'` yet, still having the same issue, access denied.

**Comment by gawel** [comment]
I have a similar error following https://github.com/frappe/frappe_docker/blob/develop/docs/single-bench.md

First a lot of errors with frappe user:
 `mariadb_1                 | 2021-05-25 14:26:33 13 [Warning] Access denied for user 'frappe'@'172.24.0.11' (using password: NO)
`

Then with a weird named user:
`mariadb_1                 | 2021-05-25 14:27:34 179 [Warning] Access denied for user '_69928f0f6369849a'@'172.24.0.10' (using password: YES)
`

I've tried to start mariadb first to ensure it's started before everything else with `docker-compose -p erp up mariadb` then in another term `docker-compose -p erp up`. Same result/errors.

**Comment by CrazyHatish** [comment]
I was also having this issue, and managed to track it down to the fact that WSL reads files in Windows as having global permissions (`-rwxrwxrwx`), so if you are using Docker Desktop on Windows 10, this might be a solution.

Upon booting, the mariadb container gives the following message:
```
Warning: World-writable config file '/etc/mysql/conf.d/frappe.cnf' is ignored
```

So what I did was:

1. Shutdown and delete the docker-compose project I had created for frappe;
2. Run `docker volume prune`, since it was a clean install and I had no data to keep (it might not be needed);
3. On WSL, `cd` to the path where you cloned the repo and `cd installation` + `chmod a-wx frappe-mariadb.cnf` (I have not tested other permission combinations);
4. Run `docker-compose --project-name <project-name> up -d` again.

If the warning isn't present on the mariadb container logs, it should now work normally.

> Windows 10 v2004
> Docker Desktop with Docker Engine v20.10.5

**Comment by revant** [comment]
https://github.com/frappe/frappe_docker/wiki/Troubleshoot

Added this

**Comment by hizbul25** [comment]
If you are running erpnext inside docker container the run: ` bench set-mariadb-host mariadb`

**Comment by zwessels** [comment]
same issue as OP. not sure what the problem is but sadly its not resolved.

**Comment by pciavald** [comment]
Hi, I'm getting the same error:

```
erpnext-site-creator      | pymysql.err.OperationalError: (1045, "Access denied for user 'root'@'172.18.0.16' (using password: YES)")
```

with the docker-compose.yml file below. I need to use a modified docker-compose file because this server already uses traefik and we need folders instead of docker volumes for the backups. Also I've already solved another issue where site-creator was not able to write in `sites`, so basically before running this docker-compose file i need to issue
```
mkdir -p sites/assets
chmod 777 sites
```

```
version: '3'

networks:
  default:
    external:
      name: traefik_default

services:

  erpnext-nginx:
    image: frappe/erpnext-nginx:edge
    container_name: erpnext-nginx
    restart: on-failure
    environment:
      - FRAPPE_PY=erpnext-python
      - FRAPPE_PY_PORT=8000
      - FRAPPE_SOCKETIO=frappe-socketio
      - SOCKETIO_PORT=9000
      - SKIP_NGINX_TEMPLATE_GENERATION=0
    labels:
      traefik.enable: true
      traefik.http.services.erpnext.loadbalancer.server.port: 80
      traefik.http.routers.erpnext.rule: Host(`erp.domain.tld`)
      traefik.http.routers.erpnext.entrypoints: https
      traefik.http.routers.erpnext.tls: true
      traefik.http.routers.erpnext.tls.certresolver: httpchallenge
    volumes:
      - ./sites:/var/www/html/sites:rw
      - ./assets:/assets:rw

  erpnext-python:
    image: frappe/erpnext-worker:edge
    restart: on-failure
    container_name: erpnext-python
    environment:
      - MARIADB_HOST=mariadb
      - REDIS_CACHE=redis-cache:6379
      - REDIS_QUEUE=redis-queue:6379
      - REDIS_SOCKETIO=redis-socketio:6379
      - SOCKETIO_PORT=9000
      - AUTO_MIGRATE=1
      - WORKER_CLASS=gthread
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw
      - ./assets:/home/frappe/frappe-bench/sites/assets:rw

  frappe-socketio:
    image: frappe/frappe-socketio:edge
    container_name: erpnext-frappe-socketio
    restart: on-failure
    depends_on:
      - redis-socketio
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw
      - ./logs:/home/frappe/frappe-bench/logs:rw

  erpnext-worker-default:
    image: frappe/erpnext-worker:edge
    container_name: erpnext-worker-default
    restart: on-failure
    command: worker
    depends_on:
      - redis-queue
      - redis-cache
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw
      - ./logs:/home/frappe/frappe-bench/logs:rw

  erpnext-worker-short:
    image: frappe/erpnext-worker:edge
    container_name: erpnext-worker-short
    restart: on-failure
    command: worker
    environment:
      - WORKER_TYPE=short
    depends_on:
      - redis-queue
      - redis-cache
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw
      - ./logs:/home/frappe/frappe-bench/logs:rw

  erpnext-worker-long:
    image: frappe/erpnext-worker:edge
    container_name: erpnext-worker-long
    restart: on-failure
    command: worker
    environment:
      - WORKER_TYPE=long
    depends_on:
      - redis-queue
      - redis-cache
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw

  erpnext-schedule:
    image: frappe/erpnext-worker:edge
    container_name: erpnext-schedule
    restart: on-failure
    command: schedule
    depends_on:
      - redis-queue
      - redis-cache
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw
      - ./logs:/home/frappe/frappe-bench/logs:rw

  redis-cache:
    image: redis:latest
    container_name: erpnext-redis-cache
    restart: on-failure
    volumes:
      - ./redis-cache:/data

  redis-queue:
    image: redis:latest
    container_name: erpnext-redis-queue
    restart: on-failure
    volumes:
      - ./redis-queue:/data

  redis-socketio:
    image: redis:latest
    container_name: erpnext-redis-socketio
    restart: on-failure
    volumes:
      - ./redis-socketio:/data

  mariadb:
    image: mariadb:10.3
    container_name: erpnext-mariadb
    restart: on-failure
    environment:
      - MYSQL_ROOT_PASSWORD=admin
    volumes:
      - ./installation/frappe-mariadb.cnf:/etc/mysql/conf.d/frappe.cnf
      - ./mariadb:/var/lib/mysql

  site-creator:
    image: frappe/erpnext-worker:edge
    container_name: erpnext-site-creator
    restart: "no"
    command: new
    depends_on:
      - erpnext-python
      - mariadb
    environment:
      - SITE_NAME=erp.vsprojects.fr
      - DB_ROOT_USER=root
      - MYSQL_ROOT_PASSWORD=admin
      - ADMIN_PASSWORD=admin
      - INSTALL_APPS=erpnext
    volumes:
      - ./sites:/home/frappe/frappe-bench/sites:rw
      - ./logs:/home/frappe/frappe-bench/logs:rw
```

**Comment by pciavald** [comment]
no news on this? The provided docker-compose file should be able to reproduce the issue.

**Comment by MohammedNoureldin** [comment]
On my side I get the exact same error. I am not sure what to do to solve it..

**Comment by pciavald** [comment]
@revant can the issue be reopened as it seems it's confirmed to still exist?

**Comment by ridwankustanto** [comment]
I got the same issue, I tried to deploy frappe on production with traefik. I can access the site with https, but it said Bad Gateway. I check the mariadb, it said like this:

![image](https://user-images.githubusercontent.com/19152005/143245991-5070e243-ab2e-4437-97c1-f68ed8e48e5f.png)

**Comment by nkz-rs** [comment]
The issue still exists. Local docker (swarm), using env.local, all volumes clean, mysql/mariadb root user name is 'root', all other components get deployed (not using traefik), site-creator fails because its unable to connect to the db - error is 'access denied for user 'root', using password: yes'

**Comment by nkz-rs** [comment]
OK - got it working - the error seems to happen when site-creator is deployed from within the docker-compose.yml file - using 'docker run' from the command line works..! 

Workaround that got the installation done for me: 

- Comment out the entire site-creator service in docker-compose.yml, bring down the stack if already running
- Create an overlay network using: 'docker network create --driver overlay --attachable <NW_NAME>'
- Specify the above network in the docker-compose.yml file: 
~~~
networks:
  default:
    external: true
    name: <NW_NAME>
~~~

- Bring up the stack without site-creator
- Run the site-creator from the command line using 'docker run' - full command below:
~~~
docker run \
    -e "SITE_NAME=$SITE_NAME" \
    -e "DB_ROOT_USER=$DB_ROOT_USER" \
    -e "MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD" \
    -e "ADMIN_PASSWORD=$ADMIN_PASSWORD" \
    -e "INSTALL_APPS=erpnext" \
    -v <project-name>_sites-vol:/home/frappe/frappe-bench/sites \
    --network ${NW_NAME} \
    frappe/erpnext-worker:$VERSION new
~~~

In the above command, NW_NAME should be the same you specified in step 1 and in docker-compose. Basic principles behind the workaround: 

- Site creator is not able to connect to mysql _from within docker-compose_, but _can connect to mysql if run through docker run_ - I still dont know the reason for this, but all that we want is to get the instance up, while wiser folks look at the root cause..
- Trying to run site creator using docker run without an _attachable network_ will fail since site creator cannot connect to the default overlay network created for the stack deployed through docker compose
- So we create an _attachable_ network manually, tell docker to use this network when deploying the stack, and then run site creator manually from the command line

Hope that helps.

**Comment by legend-zeratul** [comment]
The issue still exists. Local docker (swarm), using env.local, all volumes clean, mysql/mariadb root user name is 'root', all other components get deployed (not using traefik), site-creator fails _when deployed from within docker-compose.yml_ because its unable to connect to the db - error is 'access denied for user 'root', using password: yes'

However - **the error seems to happen when site-creator is deployed from within the docker-compose.yml file - using 'docker run' from the command line works..!**

Workaround that got the installation done for me: 

- Comment out the entire site-creator service in docker-compose.yml, bring down the stack if already running
- Create an overlay network using: 'docker network create --driver overlay --attachable <NW_NAME>'
- Specify the above network in the docker-compose.yml file: 
~~~
networks:
  default:
    external: true
    name: <NW_NAME>
~~~

- Bring up the stack without site-creator
- Run the site-creator from the command line using 'docker run' - full command below:
~~~
docker run \
    -e "SITE_NAME=$SITE_NAME" \
    -e "DB_ROOT_USER=$DB_ROOT_USER" \
    -e "MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD" \
    -e "ADMIN_PASSWORD=$ADMIN_PASSWORD" \
    -e "INSTALL_APPS=erpnext" \
    -v <project-name>_sites-vol:/home/frappe/frappe-bench/sites \
    --network ${NW_NAME} \
    frappe/erpnext-worker:$VERSION new
~~~

In the above command, NW_NAME should be the same you specified in step 1 and in docker-compose. 

Basic principles behind the workaround: 

- Site creator is not able to connect to mysql _from within docker-compose_, but _can connect to mysql if run through docker run_ - I still dont know the reason for this, but all that we want is to get the instance up, while wiser folks look at the root cause..
- Trying to run site creator using docker run without an _attachable network_ will fail since site creator cannot connect to the default overlay network created for the stack deployed through docker compose
- So we create an _attachable_ network manually, tell docker to use this network when deploying the stack, and then run site creator manually from the command line

Hope that helps.

**Comment by legend-zeratul** [comment]
Once the command succeeds (site is created), finish the first time install setup, then you can comment out/remove the network - this is needed only for the brief time we run site-creator from the command line.

**Comment by kaangoksal** [comment]
are you all trying to run the docker compose file inside the frappe_docker repository? There are some files needed in the repository for the docker-compose project to work. 

```
  mariadb:
    image: mariadb:10.3
    restart: on-failure
    environment:
      - MYSQL_ROOT_PASSWORD=${MYSQL_ROOT_PASSWORD}
    volumes:
      - ./installation/frappe-mariadb.cnf:/etc/mysql/conf.d/frappe.cnf
      - mariadb-vol:/var/lib/mysql

 ```
 
 maria db needs its configuration file, without that the database will not initialize, site creator will fail..

**Comment by kaangoksal** [comment]
There is a new commit that solves the issue that I mentioned 

https://github.com/frappe/frappe_docker/commit/81ec8da94fc22691c158701ff6aa758cf5c3a0fa

**Comment by legend-zeratul** [comment]
> are you all trying to run the docker compose file inside the frappe_docker repository? There are some files needed in the repository for the docker-compose project to work.
> 
> ```
>   mariadb:
>     image: mariadb:10.3
>     restart: on-failure
>     environment:
>       - MYSQL_ROOT_PASSWORD=${MYSQL_ROOT_PASSWORD}
>     volumes:
>       - ./installation/frappe-mariadb.cnf:/etc/mysql/conf.d/frappe.cnf
>       - mariadb-vol:/var/lib/mysql
> ```
> 
> maria db needs its configuration file, without that the database will not initialize, site creator will fail..

The config files are present and configured correctly - the issue still happens. There also seems to be an issue with the config variable named as MYSQL_ROOT_PASSWORD vs MARIADB_ROOT_PASSWORD.

**Comment by lucasvazq** [comment]
I have no problem using `MYSQL_ROOT_PASSWORD`.

You can check if your config is OK using the next command:
```sh
docker-compose -f ./.devcontainer/docker-compose.yml config
```
If all is OK, it will print the config. Else, it will print an error message.

Also, you can fully rebuild your containers to check if that fixes the problem.
Since the only containers I use are from this project, I simply do: 
```sh
docker kill $(docker ps -q)
docker rm $(docker ps -a -q)
docker volume rm $(docker volume ls -q)

# Then, I open my workspace and execute `>Remote-Containers: Reopen in Container`
```

**Comment by johnlongland** [comment]
Good day
I am having the same issue and I tried some of the possible solutions offered here but I still get the authentication erros..

Access denied for user '_90d07cd745fbad19'@'172.18.0.13' (using password: YES)

I also followed the single bench install

I am running on a Digital Ocean Droplet 
2VCPU
4G RAM
Docker 19.03.12
Ubuntu 20.04

I am able to access the databse manually.

One thing I did notice is that, with the suggestions above, the image fro mariadb should be 10.3 , but when  I do
a docker ps , it still shows the image for mariadb as 10.6 ?

Are there perhaps any other suggestions, please?

**Comment by echostorm** [comment]
Same thing going on any luck with solving it?

**Comment by lucasvazq** [comment]
Have you tried <b>[[frappe_docker] Fixing MariaDB issues after rebuilding the container](https://github.com/frappe/frappe_docker/tree/main/development#fixing-mariadb-issues-after-rebuilding-the-container)</b> ?

**Comment by jreduar** [comment]
I found the following, when the setup is done, the mariaDB user is set to another Host (or maybe change while build) and its rejected when the main container wants interact, so I had done the following to make it work:

-  connect to the mariadb container and delete the user displayed inside the site_config.json at the main container (DROP USER)
- Create a new user with the credentials showed at the site_config.json ( CREATE USER, with the '%'  so all host fall into)
- Assign the privileges to that user for the db (GRANT)

Hope this helps

**Comment by rorar** [comment]
> Have you tried [[frappe_docker] Fixing MariaDB issues after rebuilding the container](https://github.com/frappe/frappe_docker/tree/main/development#fixing-mariadb-issues-after-rebuilding-the-container) ?

The link has been moved:
[https://github.com/frappe/frappe_docker/blob/main/docs/troubleshoot.md](https://github.com/frappe/frappe_docker/blob/main/docs/troubleshoot.md)

**Comment by sudhegan** [comment]
> I made similar experiences as @jreduar. Either the password changed or the host ip changed. Inside the Frappe-bench container:
> 
> * Connect to mysql metatable with: `mysql -h mariadb -u root -prootpass mysql`
> * Use `select * from user;` to identify user names.
> * Change password with `ALTER USER 'username'@'%' IDENTIFIED BY 'newpass';`, eg. `ALTER USER '_ab9080ba42f22828'@'%' IDENTIFIED BY 'password_from_site_config.json';`
> * If USER@HOST does not yet exist `CREATE USER '_ab9080ba42f22828'@'%' IDENTIFIED BY 'password_from_site_config.json';`
> * Delete all other USER@HOST entries with `DROP USER '_ab9080ba42f22828'@'172.20.0.4';`
> * Make sure the user has the priviliges for the database: `GRANT ALL PRIVILEGES ON database_name.* TO 'username'@'localhost';`, eg. `GRANT ALL PRIVILEGES ON _41da75e60133e6ed.* TO '_41da75e60133e6ed'@'%';`

the problem still exists as the permission for the user created by bench depends on the ip of the container, we have to change it ( rename user) to make it accessible. it will good to make it accessable by default for the subnet or all IP's as container don't have static IP

**Comment by yujiahaopro** [comment]
The problem I encountered is that there is a special symbol # in the password, which causes the password to be incomplete and commented later. I replaced # with% 23 in the environment variable using URL encoding and then decoded it in the code. If I am not sure if this is the reason, I can print the password or account name, making sure to print it in the code
